### PR TITLE
DOC: sparse.linalg: Enhance lobpcg documentation with updated examples

### LINKED ISSUE #18881: DOC: lobpcg examples improvement
### Issue with current documentation:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lobpcg.html Examples includes
from scipy.sparse.linalg import LinearOperator, aslinearoperator
which is no longer use in the current development version.

### Idea or request for content:

Either remove the unused import or better yet include examples that use it

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by anudeepadi**
- Added examples utilizing `aslinearoperator` and `LinearOperator`

@lobpcg 

#### Reference issue

closes https://github.com/scipy/scipy/issues/18881

### CODE REVIEW

**Comment by lobpcg** [comment]
@anudeepadi Reference https://github.com/scipy/scipy/issues/18881

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 1]
```diff
@@ -370,6 +370,52 @@ def lobpcg(
...
     array([100.])
 
+
+    The next two examples demonstrate alternative ways to provide the matrix `A`
+    to lobpcg without constructing the full matrix:
+
+    ### Example using `aslinearoperator`:
+
+    This converts the sparse matrix `A` to a LinearOperator. The matrix `A`
+    does not need to be explicitly constructed.
```
Comment (lobpcg) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:380]: The matrix does need to be constructed to use `aslinearoperator`

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 2]
```diff
@@ -370,6 +370,52 @@ def lobpcg(
...
+    Here we define custom `matvec` and `matmat` functions for the matrix-vector and
+    matrix-matrix products. The `LinearOperator` abstracts away the matrix 
+    representation. 
+
+    >>> from scipy.sparse.linalg import lobpcg, LinearOperator
+    >>> import numpy as np
+    >>> n = 100
+    >>> vals = np.arange(1, n+1, dtype=np.float64)
+    # Define matvec for matrix-vector multiplication
+    >>> def matvec(X):
```
Comment (lobpcg) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:405]: remove. you already have the same function below called matmat

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 3]
```diff
@@ -370,6 +370,52 @@ def lobpcg(
...
+    >>>     return vals[:, np.newaxis] * X
+    # Define matmat for matrix-matrix multiplication (important for LOBPCG performance)
+    >>> def matmat(X):
+    >>>     return vals[:, np.newaxis] * X
```
Comment (lobpcg) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:409]: That is probably not a proper format for docstrings and will result in an error, I guess. Look at other examples of multi-line and edit yours accordingly

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 4]
```diff
@@ -370,6 +370,52 @@ def lobpcg(
...
+    # Create a LinearOperator with matvec and matmat
+    >>> A = LinearOperator((n, n), matvec=matmat, dtype='float64')
```
Comment (lobpcg) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:411]: You must specify both `matvec=matmat, matmat=matmat` for `lobpcg` to take advantage of `matmat`

**Comment by lobpcg** [comment]
Look at the docsting text above your insert for duplicates

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 5]
```diff
@@ -370,6 +370,43 @@ def lobpcg(
     >>> eigenvalues
     array([100.])
 
+
+    The next two examples demonstrate alternative ways to provide the matrix `A`
+    to lobpcg without constructing the full matrix:
+
+    Example using `aslinearoperator`:
+
+    This converts the sparse matrix `A` to a LinearOperator. The matrix `A` needs
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:379]:
```suggestion
    This converts the sparse matrix `A` to a `LinearOperator`. The matrix `A` needs 
```

**Review by j-bowhay — CHANGES_REQUESTED** [review]

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 6]
```diff
@@ -370,6 +370,36 @@ def lobpcg(
...
+    >>> A_aslo = aslinearoperator(A) 
+    >>> X = np.random.default_rng().normal(size=(n, 3))
+    >>> eigenvalues, _ = lobpcg(A_aslo, X, largest=False, maxiter=80)
+    >>> print(eigenvalues)  
+    [1. 2. 3.]
+
+    Example using `LinearOperator` with a callable:
+    
+    Define matmat for matrix-matrix multiplication (important for LOBPCG performance)
+    >>> def matmat(X):
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:393]: Missing whitespace
```suggestion
    Define matmat for matrix-matrix multiplication (important for LOBPCG performance)
    
    >>> def matmat(X):
```

File: scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py [hunk 7]
```diff
@@ -370,6 +370,36 @@ def lobpcg(
...
     array([100.])
 
+
+    The next two examples demonstrate alternative ways to provide the matrix `A`
+    to `lobpcg` without constructing the full matrix:
+
+    Example using `aslinearoperator`:
+
+    This converts the sparse matrix `A` to a `LinearOperator`. The matrix `A` needs 
+    to be constructed to use the `aslinearoperator`.
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py:380]: Potentially redundant information
```suggestion
```

**Review by j-bowhay — CHANGES_REQUESTED** [review]
@lobpcg did you have any further feedback?

**Comment by lobpcg** [comment]
@anudeepadi could you please add me as a collaborator so that I can also edit the file? I have a few comments, but it would be easier for me to just edit rather than explain.

**Comment by lobpcg** [comment]
I have better incorporated the new examples into the previous text. LGTM

**Review by j-bowhay — APPROVED** [review]
Thanks both

**Pull request merged**